Moving OC card method logic to the new OC payment method class

wjrosa opened the pull request:
<!--
Did I add a title? A descriptive, yet concise, title.
-->

<!--
Issue: Link to the GitHub issue this PR addresses (if appropriate).
-->

Fixes STRIPE-667

## Changes proposed in this Pull Request:

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for the PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

This PR moves all the specific Optimized Checkout logic of the card payment method to a new payment method class. The goal is to improve the code's organization and properly separate its concerns.

## Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Please follow the following guidelines when writing testing instructions:

- Include screenshots if there is no similar flow in the critical flows: https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Critical-flows
- Assume instructions will be copied over to the Release Testing Instructions wiki page: https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions
- Assume instructions will be followed by external testers.
- Assume tester does not have intimate knowledge of Stripe.
-->

- Checkout to this branch on your test environment (`dev/moving-oc-card-method-logic-to-single-method-class`)
- Connect your Stripe account
- Enable the Optimized Checkout
- Enable some payment methods
- As a shopper, add any product to your cart
- Go to the block checkout page
- Confirm the Stripe (single) payment element is displayed and you are able to complete the purchase
- Confirm the same for the classic checkout
- Confirm that you can also add payment methods

---

-   [x] Covered with tests (or have a good reason not to test in description ☝️)
-   [ ] Tested on mobile (or does not apply)

### Changelog entry

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

**Post merge**

-   [ ] Added testing instructions to the [Release Testing Instructions wiki page](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions) (or does not apply)
-   [ ] Added the [needs docs label](https://github.com/woocommerce/woocommerce-gateway-stripe/labels?q=docs) (or does not apply)
-   [ ] Included this PR in the Release Thread scope (or does not apply)

wjrosa committed:
Removing card requirement for OC

wjrosa committed:
Merge branch 'develop' into fix/removing-card-requirement-for-oc

wjrosa committed:
Backend changes

wjrosa committed:
Merge branch 'fix/removing-card-requirement-for-oc' of https://github.com/woocommerce/woocommerce-gateway-stripe into fix/removing-card-requirement-for-oc

wjrosa committed:
Backend changes

wjrosa committed:
Changelog and readme entries

wjrosa committed:
Fix tests

wjrosa committed:
Fix tests

wjrosa committed:
unit test

wjrosa committed:
Merge branch 'develop' into fix/removing-card-requirement-for-oc

wjrosa committed:
Fix tests

wjrosa committed:
Fix shortcode checkout methods display

wjrosa committed:
Moving OC card method logic to the new single method class

wjrosa committed:
Removing unnecessary condition

wjrosa committed:
Condition is actually necessary

wjrosa committed:
Merge branch 'fix/removing-card-requirement-for-oc' into dev/moving-oc-card-method-logic-to-single-method-class

wjrosa committed:
Fix shortcode container label

wjrosa committed:
Moving some methods

wjrosa committed:
Additional fixes

wjrosa committed:
Fix add payment method title

wjrosa committed:
Adding specific frontend constant to avoid confusion

wjrosa committed:
Renaming class to OC

wjrosa committed:
Improving docs

wjrosa committed:
Merge branch 'develop' into fix/removing-card-requirement-for-oc

wjrosa committed:
Removing disabledButChecked prop and related code

wjrosa committed:
Removing the required for OC pill and related files

wjrosa committed:
Renaming Single to OC

wjrosa committed:
Fix tests

wjrosa committed:
Unit test

wjrosa committed:
Changelog and readme entries

Mayisha reviewed (commented):
@wjrosa I am getting some errors when testing this branch. 

**Settings**
✅ Settings page looks good.

**Shortcode checkout**
✅ Single payment element is displayed on the checkout page.
<img width="486" height="708" alt="Screenshot 2025-08-14 at 5 09 43 PM" src="https://github.com/user-attachments/assets/fcda1458-5201-4c8c-ba85-4966c56c7255" />
🔴 Checkout fails for me with the following error-

<img width="1111" height="132" alt="Screenshot 2025-08-14 at 5 14 44 PM" src="https://github.com/user-attachments/assets/de56d548-c588-44d5-ab41-2dc0c3d6c133" />

And I see this in my logs

<img width="1259" height="484" alt="Screenshot 2025-08-14 at 4 53 10 PM" src="https://github.com/user-attachments/assets/ccf184a8-b0c4-440c-bc1c-7e8e86325b58" />


**Block checkout**
⚠️ Single payment element is displayed on the block checkout page. Probably unrelated to this PR but I don't see the test instructions on the block checkout page. But I see them in shortcode.

<img width="694" height="686" alt="Screenshot 2025-08-14 at 5 09 19 PM" src="https://github.com/user-attachments/assets/938f345f-0358-4803-acd1-53cafa565777" />

🔴 Checkout fails with the following error-

<img width="1051" height="123" alt="Screenshot 2025-08-14 at 4 59 11 PM" src="https://github.com/user-attachments/assets/bc70ade1-24a8-4a17-870c-4175c0405ae3" />

wjrosa committed:
Merge branch 'develop' into fix/removing-card-requirement-for-oc

wjrosa committed:
Merge branch 'dev/removing-required-for-oc-pill' into fix/removing-card-requirement-for-oc

wjrosa committed:
Hide OC container when there are no methods available

wjrosa committed:
Merge branch 'develop' into fix/removing-card-requirement-for-oc

wjrosa committed:
Merge branch 'fix/removing-card-requirement-for-oc' into dev/moving-oc-card-method-logic-to-single-method-class

wjrosa committed:
Merge branch 'develop' into fix/removing-card-requirement-for-oc

wjrosa committed:
Merge branch 'develop' into fix/removing-card-requirement-for-oc

wjrosa committed:
Fix merge

wjrosa committed:
Fix tests

wjrosa committed:
Merge branch 'develop' into fix/removing-card-requirement-for-oc

wjrosa committed:
Merge branch 'fix/removing-card-requirement-for-oc' into dev/moving-oc-card-method-logic-to-single-method-class

wjrosa committed:
Merge branch 'develop' into fix/removing-card-requirement-for-oc

wjrosa committed:
Fix express methods not available

wjrosa committed:
Merge branch 'fix/removing-card-requirement-for-oc' into dev/moving-oc-card-method-logic-to-single-method-class

wjrosa committed:
Removing unnecessary import

wjrosa committed:
Fix payment method unavailable

wjrosa committed:
Fix payment method unavailable

wjrosa committed:
Fix method ID

wjrosa committed:
Removing unnecessary changes

wjrosa committed:
Revert unnecessary changes

wjrosa committed:
Simplifying logic

wjrosa committed:
Simplifying logic

wjrosa committed:
Fix payment method title

wjrosa committed:
Revert unnecessary change

wjrosa committed:
Flagging infinity loop line

wjrosa committed:
Fix infinity loop

wjrosa committed:
Removing unnecessary constant

wjrosa committed:
Simplifying logic

wjrosa committed:
Fix instructions

wjrosa committed:
Improving readability

wjrosa committed:
Improving readability

wjrosa committed:
Improving readability

wjrosa commented:
Hey @Mayisha! Thanks for the review. I identified several issues with this PR after your review. I have updated it with the base branch and fixed all the bugs I could find (including the ones you pointed out). Could you please give it another round of review? Thank you 🙏

Mayisha commented on changelog.txt:
```diff
@@ -1,6 +1,8 @@
 *** Changelog ***
 
 = 9.9.0 - xxxx-xx-xx =
+* Dev - Moves all the Optimized Checkout feature code of the card payment method class to a new, specific class
```
```suggestion
* * Dev - Extracts Optimized Checkout logic from card payment method into new WC_Stripe_UPE_Payment_Method_OC class
```
Instead of 'a new, specific class', naming the class might be better.

Mayisha reviewed (commented):
@wjrosa card payment is working on both block and shortcode checkout pages but I can not pay with any other redirect methods. I have tested with EPS and iDeal. Checkout fails for both with the following error-

<img width="1272" height="280" alt="Screenshot 2025-08-20 at 3 56 09 PM" src="https://github.com/user-attachments/assets/96823c47-4226-46c5-9eac-09cde7506690" />

Could you confirm if it is related or unrelated to this PR?

Mayisha commented:
@wjrosa I am noting two more issues here. Please feel free to create separate issues to handle them if they are unrelated to the changes here.

**Steps to reproduce**
- Enable OC
- Enable card and another method
- Enable Google/Apple Pay, Link
- Save the settings
- Go to the checkout page. Payment methods are displayed correctly ✅ 
- Disable card and the other method.
- The express checkout section gets hidden with a message (this is expected) ✅ 

<img width="1120" height="163" alt="Screenshot 2025-08-20 at 4 14 19 PM" src="https://github.com/user-attachments/assets/fe2bc9ee-8392-47e9-a345-94225d01fb14" />

- Now go to the checkout page. There are 2 issues
1. The payment element loads with an error.

<img width="793" height="338" alt="Screenshot 2025-08-20 at 4 12 40 PM" src="https://github.com/user-attachments/assets/2a152835-6cf2-4828-bd6a-912d5e60963b" />


2. The Link button is present in the express payment section (this is not present when OC is disabled)

<img width="1199" height="156" alt="Screenshot 2025-08-20 at 4 14 38 PM" src="https://github.com/user-attachments/assets/bf5727a2-e1f4-4d12-a722-9ab47915b3ba" />

wjrosa committed:
Merge branch 'develop' into dev/moving-oc-card-method-logic-to-single-method-class

wjrosa committed:
Update changelog.txt

Co-authored-by: Mayisha <33387139+Mayisha@users.noreply.github.com>

wjrosa committed:
Changelog and readme entries

wjrosa committed:
Merge branch 'develop' into dev/moving-oc-card-method-logic-to-single-method-class

wjrosa committed:
Merge branch 'develop' into dev/moving-oc-card-method-logic-to-single-method-class

wjrosa committed:
Merge branch 'dev/moving-oc-card-method-logic-to-single-method-class' of https://github.com/woocommerce/woocommerce-gateway-stripe into dev/moving-oc-card-method-logic-to-single-method-class

wjrosa committed:
Fix invalid intent log when intent data is missing

wjrosa committed:
Fix selected payment method type

wjrosa commented:
> @wjrosa card payment is working on both block and shortcode checkout pages but I can not pay with any other redirect methods. I have tested with EPS and iDeal. Checkout fails for both with the following error-

Good catch! That was a new validation we recently added for the payment intent. I was able to reproduce it here and fixed it in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/4579/commits/ce4817a58bca77fd5b2190498f7cb2530350e9f1.

wjrosa committed:
Revert order lock changes

wjrosa committed:
Fix methods availability

wjrosa commented:
> The payment element loads with an error.

Thanks for the steps to reproduce the issue! I haved fixed it in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/4579/commits/c28b90330f7bdff048d1853f6e8066720019663b

> The Link button is present in the express payment section (this is not present when OC is disabled)

This specific behavior will be handled in STRIPE-678

wjrosa committed:
Merge branch 'develop' into dev/moving-oc-card-method-logic-to-single-method-class

Mayisha commented on includes/class-wc-stripe-helper.php:
```diff
@@ -1949,11 +1949,13 @@ public static function validate_intent_for_order( $order, $intent, ?string $sele
 		$order_amount          = WC_Stripe_Helper::get_stripe_amount( $order->get_total(), $order->get_currency() );
 		$order_intent_id       = self::get_intent_id_from_order( $order );
+		$intent_currency       = isset( $intent->currency ) ? strtolower( $intent->currency ) : null;
+		$intent_amount         = isset( $intent->amount ) ? (int) $intent->amount : null;
```
```suggestion
		$intent_amount         = isset( $intent->amount ) ? (float) $intent->amount : null;
```

Mayisha commented on includes/payment-methods/class-wc-stripe-upe-payment-method-oc.php:
```diff
@@ -0,0 +1,115 @@
+
+	const STRIPE_ID = WC_Stripe_Payment_Methods::OC;
+
+	const LPM_GATEWAY_CLASS = WC_Gateway_Stripe::class;
```
Nit: I think we don't need `LPM_GATEWAY_CLASS`. It was originally meant for mapping between UPE and legacy methods.

wjrosa replied:
Good to know. Removed it in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/4579/commits/f95015e4140bc2e4b2f643d8862f14c83f8d44ba

Mayisha commented on includes/payment-methods/class-wc-stripe-upe-payment-method-oc.php:
```diff
@@ -0,0 +1,115 @@
+	}
+
+	/**
+	 * The Credit Card method allows automatic capture.
```
Quick check: is this the right docblock for this method?

wjrosa replied:
Not really. Thanks for flagging it! Fixed all of them in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/4579/commits/bf66fb12d0d4b75abf8a019233718026752be3b4

Mayisha reviewed (changes requested):
Thanks for the updates @wjrosa. The result of functional testing with OC enabled is as following-

✅ Settings looks good.
✅ Enabled methods are rendered correctly on shortcode checkout page.
✅ Enabled methods are rendered correctly on block checkout page
✅ Card payment is working on both checkout pages.
✅ Redirect payment is working on both checkout pages.
🔴 When all the payment methods are disabled I see an error on shortcode checkout page. Block checkout looks good.
<img width="513" height="415" alt="Screenshot 2025-08-25 at 6 20 45 PM" src="https://github.com/user-attachments/assets/0e0b9b86-c62d-4ff5-aed7-88fc3395d945" />
🔴 Payment method name is wrong when I pay with Google Pay

|page|`develop`|this branch|
|-----|----------|------------|
|Order received page| <img width="1068" height="442" alt="Screenshot 2025-08-25 at 7 22 35 PM" src="https://github.com/user-attachments/assets/f57fc489-eb86-47da-988d-4aca66afbae3" /> | <img width="1068" height="442" alt="Screenshot 2025-08-25 at 7 25 01 PM" src="https://github.com/user-attachments/assets/8c04fab2-0479-482e-bd83-2592a6b05466" /> |
|Order edit page| <img width="1000" height="156" alt="Screenshot 2025-08-25 at 7 25 40 PM" src="https://github.com/user-attachments/assets/4d035bc3-e308-4e32-b810-aa548ed4a36d" /> | <img width="1000" height="156" alt="Screenshot 2025-08-25 at 7 25 45 PM" src="https://github.com/user-attachments/assets/f1abd2ff-97bb-47d7-b90b-32f450d580af" /> |

wjrosa committed:
Update includes/class-wc-stripe-helper.php

Co-authored-by: Mayisha <33387139+Mayisha@users.noreply.github.com>

wjrosa committed:
Removing unnecessary LPM_GATEWAY_CLASS constnat

wjrosa committed:
Update doc blocks

wjrosa committed:
Merge branch 'develop' into dev/moving-oc-card-method-logic-to-single-method-class

wjrosa committed:
Fix methods availability at the shortcode checkout

wjrosa committed:
Merge branch 'dev/moving-oc-card-method-logic-to-single-method-class' of https://github.com/woocommerce/woocommerce-gateway-stripe into dev/moving-oc-card-method-logic-to-single-method-class

wjrosa committed:
Fix ECE methods title

wjrosa commented:
Thanks for the great review once again, @Mayisha! And for the patience here.

> 🔴 When all the payment methods are disabled I see an error on shortcode checkout page. Block checkout looks good.

Fixed it in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/4579/commits/7feb8d98fb0f9c5309bd35c0a5f3a8258501ccf1

> 🔴 Payment method name is wrong when I pay with Google Pay

Fixed it in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/4579/commits/4d60f51f7b21f72e5357ed7a8bcdd64767317464

wjrosa committed:
Fix payment intent amount comparison

Mayisha reviewed (approved):
Thanks @wjrosa for the fixes. Confirming that I don't see any of the errors/issues I mentioned in my previous comments.

**Settings**
✅ OC option is rendered correctly irrespective of the enabled payment methods.

**Block checkout**
✅ OC element is rendered when there are enabled payment methods.
✅ OC element is not rendered when there are no enabled payment methods. The express methods are also hidden.
✅ Checkout with card is successful.
✅ Checkout with Google Pay is successful and the payment method title is correct.
✅ Checkout with redirect payment method (iDeal/EPS) is successful.

**Shortcode checkout**
✅ OC element is rendered when there are enabled payment methods.
✅ OC element is not rendered when there are no enabled payment methods. The express methods are also hidden.
✅ Checkout with card is successful.
✅ Checkout with Google Pay is successful and the payment method title is correct.
✅ Checkout with redirect payment method (iDeal/EPS) is successful.

⚠️ Noting one observation with Link as I see [STRIPE-678](https://linear.app/a8c/issue/STRIPE-678) is merged. When I have no payment methods enabled, I now see the Link button is available in the settings. But I don't see the Link button on checkout page in this branch.

<img width="769" height="243" alt="Screenshot 2025-08-26 at 12 06 25 PM" src="https://github.com/user-attachments/assets/2c4d3cf4-ebc7-4bad-a242-c28b2219d64e" />

malithsen commented on includes/constants/class-wc-stripe-payment-methods.php:
```diff
@@ -33,6 +33,7 @@ class WC_Stripe_Payment_Methods {
 	const SEPA_DEBIT        = 'sepa_debit';
 	const SOFORT            = 'sofort';
 	const WECHAT_PAY        = 'wechat_pay';
+	const OC                = 'card'; // This is a special case for the Optimized Checkout
```
nit: To avoid any ambiguity, should we set this to its own ID (eg: `oc_container`) and then map it to `card` where frontend/Stripe expects it?

wjrosa replied:
I tried that, but it caused a lot of bugs (tried it again now and still couldn't do it). We still need some stuff from the card payment method class (even when inactive). That was the way I could make it as independent as possible. We can revisit this later

wjrosa committed:
Fix Link availability in the block checkout

wjrosa committed:
UI update

wjrosa commented:
Thanks, Mayisha! I have fixed the availability for the block checkout and created STRIPE-687 to address the same for the shortcode checkout.

malithsen commented on includes/payment-methods/class-wc-stripe-upe-payment-method-oc.php:
```diff
@@ -0,0 +1,119 @@
+
+		// Block checkout and pay for order (checkout) page.
+		if ( ( has_block( 'woocommerce/checkout' ) || ! empty( $_GET['pay_for_order'] ) ) && ! is_wc_endpoint_url( 'order-received' ) ) { // phpcs:ignore WordPress.Security.NonceVerification
+			return 'Stripe';
```
```suggestion
            return __( 'Stripe', 'woocommerce-gateway-stripe' );
```

Should this be translated? I see we translate other payment method titles.

wjrosa replied:
It was previously translated, but since this is a company name, we decided to remove it and keep it in its original form.

malithsen commented on includes/payment-methods/class-wc-stripe-upe-payment-method-oc.php:
```diff
@@ -0,0 +1,119 @@
+	 * @inheritDoc
+	 */
+	public function is_available() {
+		return true;
```
Should we do something like

```suggestion
    if ( ! parent::is_available() ) return false;
    return true; 
```

I see the `is_available` method in `WC_Payment_Gateway` (which was in the previous inheritance chain) has some additional checks that might be useful.

wjrosa replied:
Sure. Added it in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/4579/commits/f58a3fea252c55953956ff40efa337851e73803a. I didn't apply your suggestion directly due to linting.

malithsen reviewed (approved):
Looks good to me. I left some comments, but they are not merge blockers.

wjrosa committed:
Adding call to parent is_available method

wjrosa committed:
Fix tests
